[WIP] Add foreman_anislbe_roles module

branic opened the pull request:
This PR adds a foreman_ansible_roles module.

This module will allow working with Foreman Ansible Roles.

Marking this PR as WIP, but I wanted to get it open to get feedback on the current (although limited) functionality.

evgeni commented:
Is state=fetch used to list the available roles on a proxy?

If so, I'd prefer if we could have one module for listing, and one for performing actions, as that's what current ansible best practice is.

branic commented:
> Is state=fetch used to list the available roles on a proxy?
> 
> If so, I'd prefer if we could have one module for listing, and one for performing actions, as that's what current ansible best practice is.

@evgeni The fetch action lists the Ansible Roles that have been placed in `/etc/ansible/roles` and are available to be imported.

I don't mind making that action a separate module if that's the desired direction. I think the `foreman_search_facts` module will be able to be used for listing roles that have already been imported.

Would you like fetch to be its own module?

branic commented on plugins/modules/foreman_ansible_roles.py:
```diff
@@ -0,0 +1,100 @@
+    module = ForemanAnsibleRolesModule(
+        foreman_spec = dict(
+            state = dict(default='fetch', choices=['fetch']),
+            proxy_id = dict(),
```
Thanks for the suggestion @mdellweg . I'll make the change.

branic replied:
@mdellweg When implementing either one of the suggestions I'm getting the below error:

```
{"msg": "Unsupported parameters for (basic.py) module: proxy_id Supported parameters include: password, proxy, server_url, state, username, validate_certs", "failed": true, "invocation": {"module_args": {"username": "admin", "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "server_url": "https://satellite.example.com", "validate_certs": "false", "state": "fetch", "proxy_id": "1"}}}
SystemExit: 1
```

The only difference in the error message between the two suggestions is that `smart_proxy` or `proxy` is listed in the `Supported parameters` list in the error message. The proxy_id is appearing in the `module_args` as expected.

Any ideas on how to fix the error are appreciated.

mdellweg replied:
Well i should have explained better. In order to be able to write playbooks that cannot know any foreman internal database id's, our modules describe all entities (including smart_proxies) by their name. So with either of the suggestions, it is expected that the module only accepts the proxy in the (smart_)proxy field, and you should specify it's fqdn there. That is why the module than (in turn) needs to do some lookup to figure out its id at runtime. Have a look at (i.e.) the domain module.

branic replied:
It looked to me like using `entity` type syntax you provided would automatically resolve to the ID, but had to use `find_resource_by_name` to find the id that needed to be provided to the `fetch` API call.

evgeni replied:
The automatic resolution only happens if you call `module.run()` or `_entity, module_params = module.resolve_entities(module_params)`. The later is probably what you're looking for here.

branic replied:
Thanks @evgeni. I've modified the code to call `resolve_entities`.

branic commented:
I've renamed the module to `foreman_anisble_roles_facts` and implemented not only `fetch` but also `list` and `show`.

I'll limit this PR to just the fact gathering module and open another PR to implement the `import`, `delete`, and `obsolete`.

I'll start working on adding test cases.

branic commented:
@evgeni What is the recommended way to get files on the filesystem when running tests? 

I need to place a test role in `/etc/ansible/roles` so that fetch will have data to report. I'll need this in the future for testing import functionality as well.

branic commented:
@evgeni One other question. During the test run I'm creating the Organization and Location, but I need to assign the location to the smart proxy. I'm not finding any functionality to do that. Is there functionality to do that?

evgeni commented:
> @evgeni One other question. During the test run I'm creating the Organization and Location, but I need to assign the location to the smart proxy. I'm not finding any functionality to do that. Is there functionality to do that?

Nope, sorry. We have https://github.com/theforeman/foreman-ansible-modules/issues/149 open, and I think there even was a PR for it, but never got merged :(

evgeni commented:
> @evgeni What is the recommended way to get files on the filesystem when running tests?
> 
> I need to place a test role in `/etc/ansible/roles` so that fetch will have data to report. I'll need this in the future for testing import functionality as well.

Sadly not. We have the same issue with Puppet modules too. In the past, I've been using the following snippet
https://github.com/theforeman/foreman-ansible-modules/blob/41d0a7aa3f35bff406af9e90e3cf68b08f21ba01/tests/test_playbooks/host.yml#L2-L10

and adding a `foreman` host to the inventory that is my real Foreman installation, not nice, but at least documented. you can also do it by hand and write it in the header of the test playbook:

https://github.com/theforeman/foreman-ansible-modules/blob/41d0a7aa3f35bff406af9e90e3cf68b08f21ba01/tests/test_playbooks/hostgroup.yml#L2-L8

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+            action = dict(default='fetch', choices=['fetch', 'list', 'show']),
+            proxy = dict(type='entity', flat_name='proxy_id', aliases=['smart_proxy'], resource_type='smart_proxies'),
+            search=dict(default=""),
+            full_details=dict(type='bool', aliases=['info'], default='false'),
```
```suggestion
            full_details=dict(type='bool', aliases=['info'], default=False),
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+            proxy = dict(type='entity', flat_name='proxy_id', aliases=['smart_proxy'], resource_type='smart_proxies'),
+            search=dict(default=""),
+            full_details=dict(type='bool', aliases=['info'], default='false'),
+            organization=dict(type='entity', flat_name='organization_id', resource_type='organizations'),
+            location=dict(type='entity', flat_name='location_id', resource_type='locations'),
```
```suggestion
            organization=dict(type='entity'),
            location=dict(type='entity'),
```

our helper should autodetect these

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+
+    module = ForemanAnsibleRolesFactsModule(
+        foreman_spec = dict(
+            action = dict(default='fetch', choices=['fetch', 'list', 'show']),
+            proxy = dict(type='entity', flat_name='proxy_id', aliases=['smart_proxy'], resource_type='smart_proxies'),
```
```suggestion
            action=dict(default='fetch', choices=['fetch', 'list', 'show']),
            proxy=dict(type='entity', flat_name='proxy_id', aliases=['smart_proxy'], resource_type='smart_proxies'),
```

the linter is picky ;)

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+        module_params['full_details'] = True
+
+    with module.api_connection():
+        _entity, module_params = module.resolve_entities(module_params, 'ansible_roles')
```
```suggestion
        module.auto_lookup_entities()
        params = _flatten_entity(module.foreman_params, module.foreman_spec)
```

I'm afraid we managed to leave this PR around for too long and merged a huge refactoring since then. I'll provide suggestions how to update.

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+    with module.api_connection():
+        _entity, module_params = module.resolve_entities(module_params, 'ansible_roles')
+
+        if 'proxy' in module_params:
+            params['proxy_id'] = module_params['proxy']['id']
+
+        if 'organization' in module_params:
+            params['organization_id'] = module_params['organization']['id']
+
+        if 'location' in module_params:
+            params['location_id'] = module_params['location']['id']
```
```suggestion
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+        if 'location' in module_params:
+            params['location_id'] = module_params['location']['id']
+
+        if module_params['action'] == 'fetch':
+            resources = module.fetch_resource('ansible_roles', params)
+            module.exit_json(ansible_roles=resources['results']['ansible_roles'])
+            # module.exit_json(mod_parms=module_params, my_parms=params)
+
+        if module_params['action'] == 'list':
+            if 'organization' in module_params:
+                params['organization_id'] = module.find_resource_by_name('organizations', module_params['organization'], thin=True)['id']
+
+            response = module.list_resource(resource, search, params)
+
+            if module_params['full_details']:
+                resources = []
+                for found_resource in response:
+                    resources.append(module.show_resource(resource, found_resource['id'], params))
+            else:
+                resources = response
+
+            module.exit_json(ansible_roles=resources)
```
```suggestion
        if module.foreman_params['action'] == 'fetch':
            resources = module.fetch_resource('ansible_roles', params)
            resources = resources['results']['ansible_roles']
        elif module.foreman_params['action'] == 'list':
            response = module.list_resource(resource, module.foreman_params.get('search'), params)

            if module.foreman_params['full_details']:
                resources = []
                for found_resource in response:
                    resources.append(module.show_resource(resource, found_resource['id'], params))
            else:
                resources = response

        module.exit_json(ansible_roles=resources)
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+        ],
+    )
+
+    module_params = module.clean_params()
+    resource = 'ansible_roles'
+    search = module_params['search']
+    params = {}
```
```suggestion
    resource = 'ansible_roles'
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+  type: list
+'''
+
+from ansible.module_utils.foreman_helper import ForemanEntityAnsibleModule
+
+
+class ForemanAnsibleRolesFactsModule(ForemanEntityAnsibleModule):
+    pass
```
```suggestion
from ansible.module_utils.foreman_helper import ForemanAnsibleModule, _flatten_entity


class ForemanAnsibleRolesFactsModule(ForemanAnsibleModule):
    pass
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,186 @@
+    search = module_params['search']
+    params = {}
+
+    if module_params['action'] == 'show':
+        module_params['action'] = 'list'
+        module_params['full_details'] = True
```
```suggestion
    if module.foreman_params['action'] == 'show':
        module.foreman_params['action'] = 'list'
        module.foreman_params['full_details'] = True
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,180 @@
+        elif module.foreman_params['action'] == 'list':
+            # According to the APIDoc organization_id and location_id are valid
+            # parameters, but a "500 Internal Server Error" error is returned
+            # when an organization_id or location_id are present in the request
```
The error raised is "Association not found for organization", and I've seen that before, just can't remember where.

evgeni replied:
Ah, here: https://projects.theforeman.org/issues/25277

Will need to talk to @ares about that

evgeni replied:
I've opened https://projects.theforeman.org/issues/29583 -- could you please refer to that in the comment?

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,180 @@
+description:
+  - Gather facts about Foreman Ansible Roles
+author:
+  - "Brant Evans (@branic)
```
```suggestion
  - "Brant Evans (@branic)"
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,180 @@
+
+    module = ForemanAnsibleRolesFactsModule(
+        foreman_spec = dict(
+            action=dict(default='fetch', choices=['fetch', 'list', 'show']),
```
this contradicts with the documented default above ;)

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,180 @@
+    type: str
+  proxy:
+    description:
+      - The Smart Porxy name to use for fetching Ansible Roles
```
```suggestion
      - The Smart Proxy name to use for fetching Ansible Roles
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,180 @@
+      - The Smart Porxy name to use for fetching Ansible Roles
+    required: false
+    type: str
+    alias: smart_proxy
```
```suggestion
    aliases:
      - smart_proxy
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,180 @@
+def main():
+
+    module = ForemanAnsibleRolesFactsModule(
+        foreman_spec = dict(
```
```suggestion
        foreman_spec=dict(
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,180 @@
+
+        module.exit_json(ansible_roles=resources)
+
+if __name__ == '__main__':
```
```suggestion

if __name__ == '__main__':
```

branic replied:
Fixed.

evgeni commented on plugins/modules/foreman_ansible_roles_facts.py:
```diff
@@ -0,0 +1,180 @@
+        module.exit_json(ansible_roles=resources)
+
+if __name__ == '__main__':
+    main()
```
```suggestion
    main()

```

mdellweg replied:
Oh, let's hope, this only introduces a single line feed at the end...

evgeni replied:
Oh yeah. Newline at the end of the file is all that's needed here.

branic replied:
Fixed.

evgeni commented:
Thanks @branic! And once more, sorry for keeping you busy by catching up on changes we made to our helpers.

I've tried your module, and it works as advertised. However, I am a bit confused about the exposed workflow(s).

1. The `list`/`show` actions effectively duplicate the functionality of the `foreman_search_facts` module. I do agree that "one module per resource" is easier for the users (and would even go so far to say we should implement that for all our resources), but at the same time it feels weird to have it in the same module as `fetch` which is about a completely different dataset.
2. The `fetch` action is the really interesting feature here, as it allows us to find modules that are available for import.
3. Having an `action` parameter in a "facts" module is rather confusing, as facts are informational not actionable and IMHO exposing the API action name to the user is too much.

So if I'd be king (I am not, lucky you), I'd name this module `foreman_ansible_importable_roles_info`¹ (or `…available_roles`? or switch roles and the term? not sure) and only make it implement the `fetch` action (thus dropping the parameter altogether). And then I'd add a second module `foreman_ansible_installed_roles` (or whatever) that will again has no `action` param, and will do either list or show, depending on `full_details` or not.

¹: there has been a huge movement in Ansible to move away from the "facts" term to "info" - see https://docs.ansible.com/ansible/latest/modules/list_of_all_modules.html how all _facts modules are marked as deprecated in favor of their _info siblings

evgeni commented:
Oh, and I didn't see any difference between list and show in my tests. What extra info should show expose?

branic commented:
> Oh, and I didn't see any difference between list and show in my tests. What extra info should show expose?

I haven't seen any difference in list and show in my tests either. The API has both a list and a show endpoint which was why I implemented both.

evgeni commented:
Yeah. I think list and show are just default methods of the api and thus exist here. Show is cheaper if you only need to see one item. But in our specific case we only need the list IMHO.

branic commented:
For the workflow, I've been conflicted as well. The list/show functionality does overlap with the `foreman_search_facts` module. I implemented them in with the fetch as I was looking at the module as returning available roles that either could be imported or that already were.

I agree it does make sense to split the fetch and list/show into their own modules. As for having a dedicated list/show module is that a direction you would like the project to go or should users just rely on the `foreman_search_facts` to gather that information?

evgeni commented:
Long term I'd love to see us having separate "info" modules for the various resources, but it's far from a priority as long foreman_search_facts covers the same functionality with a slightly more complicated interface.

branic commented:
Great. I'll work on splitting the module into two modules.

Thanks for all the great feedback.

evgeni commented on plugins/modules/foreman_ansible_roles_importable_info.py:
```diff
@@ -0,0 +1,106 @@
+      - The Smart Proxy name to use for fetching Ansible Roles
+    required: false
+    type: str
+    aliases: 
```
```suggestion
    aliases:
```

(the linter is unhappy)

evgeni commented on plugins/modules/foreman_ansible_roles_importable_info.py:
```diff
@@ -0,0 +1,106 @@
+  proxy:
+    description:
+      - The Smart Proxy name to use for fetching Ansible Roles
+    required: false
```
```suggestion
    required: true
```

evgeni commented on plugins/modules/foreman_ansible_roles_importable_info.py:
```diff
@@ -0,0 +1,106 @@
+
+DOCUMENTATION = '''
+---
+module: foreman_ansible_roles_facts
```
```suggestion
module: foreman_ansible_roles_importable_info
```

evgeni commented on plugins/modules/foreman_ansible_roles_importable_info.py:
```diff
@@ -0,0 +1,106 @@
+    module = ForemanAnsibleRolesImportableInfoModule(
+        foreman_spec=dict(
+            proxy=dict(type='entity', flat_name='proxy_id', aliases=['smart_proxy'], resource_type='smart_proxies', required=True),
+            organization=dict(type='entity'),
+            location=dict(type='entity'),
```
are those two actually used for anything? I know you can pass them in the request, but I don't see any use for doing so. the data on disk (which this action lists) is not org/loc aware.

I *think* the API just allows to pass those, as it allows this on every request.

evgeni replied:
@ares is that right? 'fetch' should not be scoped?

evgeni commented on plugins/modules/foreman_ansible_roles_imported_info.py:
```diff
@@ -0,0 +1,137 @@
+
+    module = ForemanAnsibleRolesImportedInfoModule(
+        foreman_spec=dict(
+            search=dict(default=""),
```
```suggestion
            search=dict(),
```

no need for a default

evgeni commented on plugins/modules/foreman_ansible_roles_imported_info.py:
```diff
@@ -0,0 +1,137 @@
+    module = ForemanAnsibleRolesImportedInfoModule(
+        foreman_spec=dict(
+            search=dict(default=""),
+            full_details=dict(type='bool', aliases=['info'], default=False),
```
given the API does not return more info, shall we drop that param?

evgeni commented on plugins/modules/foreman_ansible_roles_imported_info.py:
```diff
@@ -0,0 +1,137 @@
+
+DOCUMENTATION = '''
+---
+module: foreman_ansible_roles_facts
```
```suggestion
module: foreman_ansible_roles_imported_info
```
